libraries/botbuilder-core/botbuilder/core/teams/teams_activity_handler.py

```diff
@@ -0,0 +1,404 @@
+                    return await self.on_teams_channel_deleted_activity(
+                        channel_data.channel, channel_data.team, turn_context
+                    )
+                if channel_data.event_type == "channelRenamed":
+                    return await self.on_teams_channel_renamed_activity(
+                        channel_data.channel, channel_data.team, turn_context
+                    )
+                if channel_data.event_type == "teamRenamed":
+                    return await self.on_teams_team_renamed_activity(
+                        channel_data.team, turn_context
+                    )
+                return await super().on_conversation_update_activity(turn_context)
+
+        return await super().on_conversation_update_activity(turn_context)
+
+    async def on_teams_channel_created_activity(  # pylint: disable=unused-argument
+        self, channel_info: ChannelInfo, team_info: TeamInfo, turn_context: TurnContext
+    ):
+        return
+
+    async def on_teams_team_renamed_activity(  # pylint: disable=unused-argument
+        self, team_info: TeamInfo, turn_context: TurnContext
+    ):
+        return
+
+    async def on_teams_members_added_dispatch_activity(  # pylint: disable=unused-argument
+        self,
+        members_added: [ChannelAccount],
+        team_info: TeamInfo,
+        turn_context: TurnContext,
+    ):
+        """
+        team_members = {}
+        team_members_added = []
+        for member in members_added:
+            if member.additional_properties != {}:
+                team_members_added.append(TeamsChannelAccount(member))
+            else:
+                if team_members == {}:
+                    result = await TeamsInfo.get_members_async(turn_context)
+                    team_members = { i.id : i for i in result }
+
+                if member.id in team_members:
+                    team_members_added.append(member)
+                else:
+                    newTeamsChannelAccount = TeamsChannelAccount(
+                        id=member.id,
+                        name = member.name,
+                        aad_object_id = member.aad_object_id,
+                        role = member.role
+                        )
+                    team_members_added.append(newTeamsChannelAccount)
+
+        return await self.on_teams_members_added_activity(teams_members_added, team_info, turn_context)
+        """
+        for member in members_added:
+            new_account_json = member.seralize()
+            del new_account_json["additional_properties"]
+            member = TeamsChannelAccount(**new_account_json)
+        return await self.on_teams_members_added_activity(members_added, turn_context)
+
+    async def on_teams_members_added_activity(
+        self, teams_members_added: [TeamsChannelAccount], turn_context: TurnContext
+    ):
+        for member in teams_members_added:
+            member = ChannelAccount(member)
```

Comment:
What is this doing?

Reply:
it's doing a downcast from TeamsChannelAccount to a normal ChannelAccount. In short Teams wanted to have us send around TeamsChannelAccount objects which are slightly richer in terms of data. When I wrote the conversation-update bot I didn't have all of the supporting classes so I wrote something to unblock myself. I'll fix this part after the PR gets in since I'll need to update the unit tests around this one as well.